### PR TITLE
Add definitions of compositeness

### LINKED ISSUE #1659: Data.Nat.Primality: add idea of Compositeness
Currently `Data.Nat.Primality` defines:

```agda
Prime : ℕ → Set
Prime 0             = ⊥
Prime 1             = ⊥
Prime (suc (suc n)) = (i : Fin n) → 2 + toℕ i ∤ 2 + n
```

This is fine, but in many situations a negative proof of `Prime n` is a lot less useful than knowing even one factor that causes it to fail the condition.

I'm proposing to add something like

```agda
Composite : ℕ → Set
Composite 0 = ⊥
Composite 1 = ⊥
Composite (suc (suc n)) = Σ[ i ∈ Fin n ] 2 + toℕ i ∣ 2 + n
```

with a corresponding

```agda
composite? : Decidable Composite
```
implemented similarly to `prime?`, along with proofs relating `Composite` and `Prime` together.

### PR DESCRIPTION
**Pull request by Taneb**
Closes #1659

### CODE REVIEW

File: src/Data/Nat/Primality.agda [hunk 1]
```diff
@@ -32,9 +47,28 @@ prime? 0             = no λ()
 prime? 1             = no λ()
 prime? (suc (suc n)) = all? (λ _ → ¬? (_ ∣? _))
 
+-- Relation between compositeness and primality
+
+composite⇒¬prime : ∀ n → Composite n → ¬ Prime n
```
Comment (MatthewDaggitt) [inline src/Data/Nat/Primality.agda:52]: Can the `n` be made implicit in these 4 proofs or does the use of `⊥` prevent that?

File: src/Data/Nat/Primality.agda [hunk 2]
```diff
@@ -10,14 +10,29 @@ module Data.Nat.Primality where
...
+Composite 0 = ⊥
+Composite 1 = ⊥
+Composite (suc (suc n)) = Σ[ i ∈ Fin n ] 2 + toℕ i ∣ 2 + n
+
+-- Decision procedure for compositeness
+
+composite? : Decidable Composite
+composite? 0 = no λ()
+composite? 1 = no λ()
+composite? (suc (suc n)) = any? λ _ → _ ∣? _
```
Comment (MatthewDaggitt) [inline src/Data/Nat/Primality.agda:34]: Minor point, but can you bracket the lambda? I know it's not necessary, but it greatly helps reading it!

**Review by MatthewDaggitt — COMMENTED** [review]
Otherwise, looks great, thanks!

File: src/Data/Nat/Primality.agda (hunk 1 above)
Comment (Taneb) [reply src/Data/Nat/Primality.agda:52]: I think it should be fine! All of these proofs have _something_ positive, either a proof of compositeness or primality or a proof of being greater than or equal to 2, that can determine n, as an argument.

File: src/Data/Nat/Primality.agda (hunk 2 above)
Comment (Taneb) [reply src/Data/Nat/Primality.agda:34]: Sure thing! I agree, it does make it a but more readable

**Pull request merged**